pkg/test/testreconciler/simpletest/controller_test.go

```diff
@@ -0,0 +1,151 @@
+				testSimpleReconciler(h, testdir, applier.Applier)
+			})
+		})
+	}
+}
+
+func testSimpleReconciler(h *testharness.Harness, testdir string, applier applier.Applier) {
+	ctx := context.Background()
+
+	k8s, err := mockkubeapiserver.NewMockKubeAPIServer(":0")
+	if err != nil {
+		h.Fatalf("error building mock kube-apiserver: %v", err)
+	}
+
+	k8s.RegisterType(schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Namespace"}, "namespaces", meta.RESTScopeRoot)
+	k8s.RegisterType(schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}, "configmaps", meta.RESTScopeNamespace)
+	k8s.RegisterType(schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Event"}, "events", meta.RESTScopeNamespace)
+	k8s.RegisterType(schema.GroupVersionKind{Group: "addons.example.org", Version: "v1alpha1", Kind: "SimpleTest"}, "simpletests", meta.RESTScopeNamespace)
+
+	defer func() {
+		if err := k8s.Stop(); err != nil {
+			h.Fatalf("error closing mock kube-apiserver: %v", err)
+		}
+	}()
+
+	addr, err := k8s.StartServing()
+	if err != nil {
+		h.Errorf("error starting mock kube-apiserver: %v", err)
+	}
+
+	klog.Infof("mock kubeapiserver will listen on %v", addr)
+
+	var requestLog httprecorder.RequestLog
+	wrapTransport := func(rt http.RoundTripper) http.RoundTripper {
+		return httprecorder.NewRecorder(rt, &requestLog)
+	}
+	restConfig := &rest.Config{
+		Host:          addr.String(),
+		WrapTransport: wrapTransport,
+	}
+
+	scheme := runtime.NewScheme()
+	if err := api.AddToScheme(scheme); err != nil {
+		h.Fatalf("error from AddToScheme: %v", err)
+	}
+
+	logger := klogr.New()
+
+	mgr, err := ctrl.NewManager(restConfig, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "",
+		Port:               0,
+		LeaderElection:     false,
+
+		// MapperProvider provides the rest mapper used to map go types to Kubernetes APIs
+		MapperProvider: restmapper.NewControllerRESTMapper,
+
+		Logger: logger,
+	})
+	if err != nil {
+		h.Fatalf("error starting manager: %v", err)
+	}
+
+	reconciler := &SimpleTestReconciler{
+		Client:  mgr.GetClient(),
+		Scheme:  mgr.GetScheme(),
+		applier: applier,
+	}
+
+	mc, err := loaders.NewManifestLoader("testdata/channels")
+	if err != nil {
+		h.Fatalf("error from loaders.NewManifestLoader: %v", err)
+	}
+	reconciler.manifestController = mc
+
+	if err = reconciler.SetupWithManager(mgr); err != nil {
+		h.Fatalf("error creating reconciler: %v", err)
+	}
+
+	if h.FileExists(filepath.Join(testdir, "before.yaml")) {
+		before := string(h.MustReadFile(filepath.Join(testdir, "before.yaml")))
+		if err := k8s.AddObjectsFromManifest(before); err != nil {
+			h.Fatalf("error precreating objects: %v", err)
+		}
+	}
+
+	mgrContext, mgrStop := context.WithCancel(ctx)
+	go func() {
+		time.Sleep(2 * time.Second)
+		// time.Sleep(200 * time.Second)
+		mgrStop()
+	}()
+	if err := mgr.Start(mgrContext); err != nil {
+		h.Fatalf("error starting manager: %v", err)
+	}
```

Comment:
Do we have to wait for one reconcile to complete and http logging to complete ?

Reply:
So it would be nice to reach a steady state, right now I'm just hard-coding a 2 second delay.  We do want to know if it enters a busy loop, so we don't want to stop after the first reconcile.  Ideally we would stop when the system became idle - no watch events, no pending reconciles; I can look at getting there but that'll be a separate PR I think!